### PR TITLE
bpo-30860: Consolidate stateful runtime globals.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
(trying again; see #2594)

* group the (stateful) runtime globals into various topical structs
* consolidate the topical structs under a single top-level _PyRuntimeState struct
* add a check-c-globals.py script that helps identify runtime globals

Other globals are excluded (see globals.txt and check-c-globals.py).

https://bugs.python.org/issue30860

### CODE REVIEW

File: Include/ceval.h [hunk 1]
```diff
@@ -93,7 +93,12 @@ PyAPI_FUNC(int) Py_GetRecursionLimit(void);
       PyThreadState_GET()->overflowed = 0;  \
     } while(0)
 PyAPI_FUNC(int) _Py_CheckRecursiveCall(const char *where);
-PyAPI_DATA(int) _Py_CheckRecursionLimit;
```
Comment (benjaminp) [inline Include/ceval.h]: This is in the stable API, no?

File: Include/internal/ceval.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#ifndef Py_INTERNAL_CEVAL_H
+#define Py_INTERNAL_CEVAL_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <ceval.h>
```
Comment (benjaminp) [inline Include/internal/ceval.h]: These sorts of includes should probably be quoted and qualified `"internal/ceval.h"`.

File: Python/pystate.c [hunk 3]
```diff
@@ -34,55 +35,66 @@ to avoid the expense of doing their own locking).
 extern "C" {
 #endif
 
-int _PyGILState_check_enabled = 1;
+void
+_PyRuntimeState_Init(_PyRuntimeState *runtime)
+{
+    _PyRuntimeState initial = {};
```
Comment (benjaminp) [inline Python/pystate.c]: Empty initializers aren't standard. You should probably just `memset` it to 0.

File: Parser/pgenmain.c [hunk 4]
```diff
@@ -21,10 +21,12 @@
 #include "node.h"
 #include "parsetok.h"
 #include "pgen.h"
+#include "internal/mem.h"
 
 int Py_DebugFlag;
 int Py_VerboseFlag;
 int Py_IgnoreEnvironmentFlag;
+struct pyruntimestate _PyRuntime = {};
```
Comment (benjaminp) [inline Parser/pgenmain.c]: ditto on empty initializer list

File: Python/pystate.c (hunk 3 above)
Comment (ericsnowcurrently) [reply Python/pystate.c]: fixed

File: Parser/pgenmain.c (hunk 4 above)
Comment (ericsnowcurrently) [reply Parser/pgenmain.c]: fixed

File: Include/internal/ceval.h (hunk 2 above)
Comment (ericsnowcurrently) [reply Include/internal/ceval.h]: fixed

File: Include/ceval.h (hunk 1 above)
Comment (ericsnowcurrently) [reply Include/ceval.h]: :(  I'm going to revert the _Py_CheckRecursionLimit part of the change for now and circle back after this patch lands.

**Review by ncoghlan — APPROVED** [review]

**Pull request merged**